filesystem/squashfs_superblock.ksy

```diff
@@ -0,0 +1,54 @@
+meta:
+  id: squashfs_superblock
+  title: SquashFS superblock
+  file-extension:
+    - squashfs
+    - snap
+  endian: le
+  license: CC-0
+doc-ref: https://github.com/AgentD/squashfs-tools-ng/blob/master/doc/format.txt
+seq:
+  - id: superblock
+    type: superblock
+types:
+  superblock:
+    seq:
+      - id: magic
+        contents: 'hsqs'
+      - id: inode_count
+        type: u4
+      - id: mod_time
+        type: u4
+      - id: block_size
+        type: u4
+      - id: frag_count
+        type: u4
+      - id: compressor
+        type: u2
+      - id: block_log
+        type: u2
+      - id: flags
+        type: u2
```

Comment:
Should probably be a bit-sized type

Reply:
[Bit mapping](http://doc.kaitai.io/user_guide.html#_bit_sized_integers) in 16-bit little endian block in Kaitai is not trivial. May need to write a script to avoid mistakes.

```
0x0001 | Inodes are stored uncompressed.            |
0x0002 | Data blocks are stored uncompressed.       |
0x0008 | Fragments are stored uncompressed.         |
0x0010 | Fragments are not used.                    |
0x0020 | Fragments are always generated.            |
0x0040 | Data has been deduplicated.                |
0x0080 | NFS export table exists.                   |
0x0100 | Xattrs are stored uncompressed.            |
0x0200 | There are no Xattrs in the archive.        |
0x0400 | Compressor options are present.            |
0x0800 | The ID table is uncompressed.
```
There are only 11 bits used, so it may seem that the mapping should start with 5 bit padding. But that's a mistake. Because the first 8 bits that Kaitai sees are from the lower byte, because the flag stored as little endian. The thing that goes first are these 7 bits.
```
0x0001 | Inodes are stored uncompressed.            |
0x0002 | Data blocks are stored uncompressed.       |
0x0008 | Fragments are stored uncompressed.         |
0x0010 | Fragments are not used.                    |
0x0020 | Fragments are always generated.            |
0x0040 | Data has been deduplicated.                |
0x0080 | NFS export table exists.                   |
```
The bits in Kaitai will be written from the end. To make it 8 bits, there should be 1 padding bit for the start of Kaitai structure. Wrong. The bit is actually 0x0004 missing in the middle. I don't know if you can compare it with spec manually, but for me it is hard.

Reply:
Done. PTAL.

Reply:
@abitrolly Note that you already can read bit integers also in little-endian byte order, see https://github.com/kaitai-io/kaitai_struct/issues/155#issuecomment-613733627. Looking at the hex values of the bits in your comment, it seems that you can read them exactly in the order as they're defined in the spec (if the whole flag is stored in little-endian bit order as you said).

Reply:
@generalmimon I don't find bit masking syntax more readable. Any attempt to work with bits without getting the byte or the word out of the flow first looks complicated. I coud use something like this.
```
seq:
  - id: flags
    type: u2
    flags:
      - pos: 0x0400
        id: someflag
      - pos: 0b00000010
        id: another
```

Reply:
@abitrolly

If the hex bit masks specified here are meant to be used on a single `u2le` parsed previously:

```
0x0001 | Inodes are stored uncompressed.            |
0x0002 | Data blocks are stored uncompressed.       |
0x0008 | Fragments are stored uncompressed.         |
0x0010 | Fragments are not used.                    |
0x0020 | Fragments are always generated.            |
0x0040 | Data has been deduplicated.                |
0x0080 | NFS export table exists.                   |
0x0100 | Xattrs are stored uncompressed.            |
0x0200 | There are no Xattrs in the archive.        |
0x0400 | Compressor options are present.            |
0x0800 | The ID table is uncompressed.
```

One can declare the structure in the `.ksy` spec in the exact same order using little-endian bit integers like this:

```yaml
types:
  flags:
    meta:
      bit-endian: le
    seq:
      - id: inodes_uncompresed # 0x0001
        type: b1
      - id: data_blocks_uncompresed # 0x0002
        type: b1
      - type: b1
        doc: 0x0004 bit is unused.
      - id: fragments_uncompresed # 0x0008
        type: b1
      - id: fragments_not_used # 0x0010
        type: b1
      - id: fragments_always_generated # 0x0020
        type: b1
      - id: data_deduplicated # 0x0040
        type: b1 
      - id: nfs_export_table # 0x0080
        type: b1

      - id: xattrs_uncompressed # 0x0100
        type: b1
      - id: xattrs_absent # 0x0200
        type: b1
      - id: compressor_options_present # 0x0400
        type: b1
      - id: id_table_uncompresed # 0x0800
        type: b1
      - type: b4 # it's recommended to omit this last unused attribute if this whole `flags` type will be used with `size: 2` everywhere
        doc: bits 11110000 are unused
```

> I don't find bit masking syntax more readable. Any attempt to work with bits without getting the byte or the word out of the flow first looks complicated.

I accept your opinion. Yeah, I realize that may be difficult to understand the pattern how the little-endian bit integers work at first glance, but once you get it, it starts to pay off with brevity, simplicity and flexibility.

And as you can see on the example above, it doesn't have to be hard to understand at all. 

It also seamlessly fits into the KS concept of declarativity, meaning that you just describe the actual bit layout of the fields, without imperatively saying how to read (or write) them. This makes serialization (kaitai-io/kaitai_struct#27) much easier as well - the bit layout is everything you need to know to be able to write the values back.

Reply:
@generalmimon is this format already implemented? I don't see `bit-endian` here http://doc.kaitai.io/ksy_reference.html 

How make sure the size is set to `u2le` for parsing this type? Maybe set it in `meta`?
```
    meta:
      bit-endian: le
      size: u2
```

Reply:
> is this format already implemented? I don't see `bit-endian` here http://doc.kaitai.io/ksy_reference.html

Yes, it is, but unfortunately the documentation for this feature hasn't been written yet, sorry for that. Hopefully I will add the section about the little-endian bit integers to the [User Guide](https://doc.kaitai.io/user_guide.html#_bit_sized_integers) soon.

I explained the bit layout of the little-endian bit-sized integers on a simple example in comment https://github.com/kaitai-io/kaitai_struct/issues/155#issuecomment-613733627, so I suggest you read it first. If you have any questions, feel free to ask on our [Gitter channel](https://gitter.im/kaitai_struct/Lobby?at=5f2d563cfe39ca5d65888894).

And please note that the KSY reference [is now disabled](https://gitter.im/kaitai_struct/Lobby?at=5f2d563cfe39ca5d65888894), refer to the [KSY syntax diagram](https://doc.kaitai.io/ksy_diagram.html) instead. I'm going to add the `bit-endian` key to the [KSY schema](https://github.com/kaitai-io/ksy_schema) soon, so it'll be available there.

> How make sure the size is set to `u2le` for parsing this type? Maybe set it in `meta`?

No, you have to make sure that every `seq` or `instance` attribute with `type: flags` has also `size: 2` (i.e. the `flags` type will be wrapped in its own substream with 2 bytes in size). Omitting the last padding field while some attribute doesn't wrap `flags` to 2-byte substream can cause wrong readings on further bit-sized integer readings using the same stream (because the internal bit position pointer `bitsLeft` stays on the wrong place in the middle of the partially-consumed byte, and just jumping out of the `flags` type doesn't reset it).

Though the compiler inserts a `alignToByte()` call if it detects a byte-aligned field (anything with `size: [0-9]+`, or `type: [us][1248]`) after a not byte-aligned bit-sized integer (`b[0-9]+`) in the same `seq`, the detection algorithm is currently quite dumb, and it doesn't recognize `if` conditions, subtypes, repetitions and type-switchings which have to be handled with special care. See https://github.com/kaitai-io/kaitai_struct/issues/743#issuecomment-623667226 for more info.

So that's why I suggest wrapping every subtype with `seq` that might not be byte-aligned in its own substream, because you're safe this way. It's sort of a workaround for this bug. But even when it's resolved, I like to make the byte `size` explicit because you immediately see how many bytes the bitfield occupies in the stream.

Reply:
Well, I'm still hoping for some response here, @abitrolly. I'd be glad if you tried the approach I presented in https://github.com/kaitai-io/kaitai_struct_formats/pull/308#discussion_r466613148, if it works for your format.

If there are any difficulties, feel free to share them, I'm here to help you.

To make it simple, you can ignore the whole thing with omitting the last padding field and adding `size`, just take the spec from https://github.com/kaitai-io/kaitai_struct_formats/pull/308#discussion_r466613148 as it is and remove my recommendation comment.

Reply:
@generalmimon [the events from](https://www.osce.org/odihr/belarus/459664) 9th of August, 2020 in Belarus were too disturbing to do coding. There is a lot of info I need to reread, and because it didn't fit in my head last time, I don't think it will fit in those spare 15 minutes I have now either. If there is already a new tutorial about bit mapping in Kaitai, I may be able to find some time to experiment with that. Right now my focus is on getting my `ksykaitai` helper released.